### PR TITLE
[Docs] Translate embed/go/memory.md to zh

### LINKED ISSUE #1091: [Docs] Translate embed/go/memory.md to zh
### The MD file link you want to translate (must be in English) 
https://github.com/WasmEdge/WasmEdge/blob/master/docs/book/en/src/embed/go/memory.md

### Target language
https://github.com/WasmEdge/WasmEdge/blob/master/docs/book/zh/src/embed/go/memory.md

### PR DESCRIPTION
**Pull request by Bevisy**
translate embed/go/memory.md to zh

Fix #1091

Signed-off-by: Binbin Zhang <binbin36520@gmail.com>

### CODE REVIEW

**Comment by Bevisy** [comment]
PTAL @alabulei1 @MileyFu

File: docs/book/zh/src/embed/go/memory.md [hunk 1]
```diff
@@ -1 +1,647 @@
-# Pass complex data
+# 将复杂参数传递给 Wasm 函数
+
+WebAssembly 规范的一个问题是它只支持非常有限的数据类型。如果要嵌入具有复杂调用参数或返回值的 WebAssembly 函数，则需要同时在 Go SDK 和 WebAssembly 函数端管理内存指针。
+此类复杂的调用参数和返回值包括字符串和字节数组等动态内存结构。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/memory.md:4]: 3 4 5 都是一行。

File: docs/book/zh/src/embed/go/memory.md [hunk 2]
```diff
@@ -1 +1,647 @@
...
+- [将字符串传递给 TinyGo 函数](#将字符串传递给-tinygo-函数)
+- [将字节传递给 Rust 函数](#将字节传递给-rust-函数)
+- [将字节传递给 TinyGo 函数](#将字节传递给-tinygo-函数)
+
+## 将字符串传递给 Rust 函数
+
+在[此示例](https://github.com/second-state/WasmEdge-go-examples/tree/master/go_MemoryGreet)中，我们将演示如何从 Go 应用程序调用[基于 Rust 的 WebAssembly 函数](https://github.com/second-state/WasmEdge-go-examples/blob/master/go_MemoryGreet/rust_memory_greet/src/lib.rs)。
+特别是，我们将讨论如何传递字符串数据。
+
+> 在 WebAssembly 中向 Rust 函数传递和返回复杂值的另一种方法是使用`wasm-bindgen`编译器工具。你可以[在这里了解更多](bindgen.md).
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/memory.md:17]: 代码块前后空格

File: docs/book/zh/src/embed/go/memory.md [hunk 3]
```diff
@@ -1 +1,647 @@
...
+
+使用标准 Rust 编译器工具将 Rust 源代码编译成 WebAssembly 字节码应用程序。
+
+```bash
+cd rust_memory_greet
+cargo build --target wasm32-wasi
+# 输出的 WASM 将是 `target/wasm32-wasi/debug/rust_memory_greet_lib.wasm`.
+```
+
+[Go SDK 应用程序](https://github.com/second-state/WasmEdge-go-examples/blob/master/go_MemoryGreet/greet_memory.go)必须从 WasmEdge 虚拟机中调用`allocate`以获得一个指向字符串参数的指针。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/memory.md:61]: 61 62 63 都是一行，然后代码块前后空格

File: docs/book/zh/src/embed/go/memory.md [hunk 4]
```diff
@@ -1 +1,647 @@
...
+Hello, WasmEdge!
+```
+
+## 将字符串传递给 TinyGo 函数
+
+在[此示例](https://github.com/second-state/WasmEdge-go-examples/tree/master/go_MemoryGreetTinyGo)中，我们将演示如何从 Go 应用程序调用[基于 TinyGo 的 WebAssembly 函数](https://github.com/second-state/WasmEdge-go-examples/blob/master/go_MemoryGreetTinyGo/greet.go)。
+
+TinyGo 函数获取字符串的内存指针，并自己构造 TinyGo 字符串。
+
+> 编译 WebAssembly 程序需要`main()`为空，才能正确设置 WASI。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/memory.md:165]: 代码块前后空格

File: docs/book/zh/src/embed/go/memory.md [hunk 5]
```diff
@@ -1 +1,647 @@
...
+  if err != nil {
+    fmt.Println("failed to load wasm")
+  }
+  vm.Validate()
+  vm.Instantiate()
+
+  subject := "WasmEdge"
+  lengthOfSubject := len(subject)
+
+  // Allocate memory for the subject, and get a pointer to it.
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/memory.md:99]: 这些注释也翻译一下吧

File: docs/book/zh/src/embed/go/memory.md [hunk 6]
```diff
@@ -1 +1,647 @@
...
+}
+```
+
+使用 TinyGo 编译器工具将 Go 源代码编译成 WebAssembly 字节码应用程序。
+
+```bash
+tinygo build -o greet.wasm -target wasi greet.go
+```
+
+[Go SDK 应用程序](https://github.com/second-state/WasmEdge-go-examples/blob/master/go_MemoryGreetTinyGo/greet_memory.go)必须从 WasmEdge 虚拟机中调用`malloc`以获得一个指向字符串参数的指针。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/memory.md:205]: 205 206 207 应该是一行吧，代码块要空格哈

File: docs/book/zh/src/embed/go/memory.md [hunk 7]
```diff
@@ -1 +1,647 @@
...
+  if err != nil {
+    fmt.Println("failed to load wasm")
+  }
+  vm.Validate()
+  vm.Instantiate()
+
+  subject := "WasmEdge"
+  lengthOfSubject := len(subject)
+
+  // Allocate memory for the subject, and get a pointer to it.
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/memory.md:243]: 注释也翻译一下

File: docs/book/zh/src/embed/go/memory.md [hunk 8]
```diff
@@ -1 +1,647 @@
...
+fibArray: [0 0 0 0 1 0 0 0 1 0 0 0 2 0 0 0 3 0 0 0 5 0 0 0 8 0 0 0 13 0 0 0 21 0 0 0 34 0 0 0]
+fib_array_return_memory memory at: 0x105430
+fibArrayReturnMemory: [0 0 0 0 1 0 0 0 1 0 0 0 2 0 0 0 3 0 0 0 5 0 0 0 8 0 0 0 13 0 0 0 21 0 0 0 34 0 0 0]
+```
+
+## 将字节传递给 TinyGo 函数
+
+在[此示例](https://github.com/second-state/WasmEdge-go-examples/tree/master/go_AccessMemoryTinyGo)中，我们将演示如何调用[基于 TinyGo 的 WebAssembly 函数](https://github.com/second-state/WasmEdge-go-examples/blob/master/go_AccessMemoryTinyGo/fib.go)以及将数组传入和传出 Go 应用程序。
+
+函数 "fibArray "接收一个数组作为调用参数，并将其填入
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/memory.md:485]: 这应该都是一行

Comment (alabulei1) [reply docs/book/zh/src/embed/go/memory.md:485]: fibArray 这里应该是代码块吧

**Review by alabulei1 — COMMENTED** [review]
Thanks for your contribution! There are two things we should take care of.
1. Add a space between code blocks.
2. Translate the noes in the code blocks.

**Comment by Bevisy** [comment]
> Thanks for your contribution! There are two things we should take care of.
> 
> 1. Add a space between code blocks.
> 2. Translate the noes in the code blocks.

@alabulei1 Thanks for your review, I have reworked the translation according to the review comments, and added all the code comments to the translation.

**Comment by alabulei1** [comment]
@Bevisy  LGTM! Only one place needs to be changed. Please check out.

**Comment by Bevisy** [comment]
> @Bevisy LGTM! Only one place needs to be changed. Please check out.

@alabulei1 Thanks, but can you mark where there are still changes to be made?

**Comment by alabulei1** [comment]
> > @Bevisy LGTM! Only one place needs to be changed. Please check out.
> 
> @alabulei1 Thanks, but can you mark where there are still changes to be made?

@Bevisy Sure. 第246行，英文字母前后空一格

**Comment by Bevisy** [comment]
> > > @Bevisy LGTM! Only one place needs to be changed. Please check out.
> > 
> > 
> > @alabulei1 Thanks, but can you mark where there are still changes to be made?
> 
> @Bevisy Sure. 第246行，英文字母前后空一格

Sorry actually forgot the `NULL` word in the comment, has been fixed now, thanks a lot.

**Comment by alabulei1** [comment]
> > > > @Bevisy LGTM! Only one place needs to be changed. Please check out.
> > > 
> > > 
> > > @alabulei1 Thanks, but can you mark where there are still changes to be made?
> > 
> > 
> > @Bevisy Sure. 第246行，英文字母前后空一格
> 
> Sorry actually forgot the `NULL` word in the comment, it's fixed now. Thanks a lot.

@Bevisy please also add a space between `NULL` in lines 237, 104, 95. Sorry for missing.

**Comment by Bevisy** [comment]
> > > > > @Bevisy LGTM! Only one place needs to be changed. Please check out.
> > > > 
> > > > 
> > > > @alabulei1 Thanks, but can you mark where there are still changes to be made?
> > > 
> > > 
> > > @Bevisy Sure. 第246行，英文字母前后空一格
> > 
> > 
> > Sorry actually forgot the `NULL` word in the comment, it's fixed now. Thanks a lot.
> 
> @Bevisy please also add a space between `NULL` in lines 237, 104, 95. Sorry for missing them.

Done 😄

**Comment by Bevisy** [comment]
@alabulei1 重新修改了标点符号，将部分句号“。”修改成冒号“：”，分别为 line 3、50、140等共13处，规范标点符号使用。
参考：https://teachinghand.com/chinese-punctuation-guide/。

PTAL @hydai

**Comment by alabulei1** [comment]
LGTM! @Bevisy Thanks for your contribution.

@hydai  Could you help merge this PR?

File: docs/book/zh/src/embed/go/memory.md [hunk 9]
```diff
@@ -1 +1,632 @@
...
+
+使用标准 Rust 编译器工具将 Rust 源代码编译成 WebAssembly 字节码应用程序：
+
+```bash
+cd rust_access_memory
+cargo build --target wasm32-wasi
+# 输出的 WASM 将是 target/wasm32-wasi/debug/rust_access_memory_lib.wasm.
+```
+
+[Go SDK 应用程序](https://github.com/second-state/WasmEdge-go-examples/blob/master/go_AccessMemory/run.go)必须从 WasmEdge 虚拟机中调用 `allocate` ，以获得一个指向数组的指针。然后它将调用Rust中的`fib_array()`函数并传入指针。在函数返回后，Go应用程序将使用 WasmEdge 的 `store` API，从调用参数（ `fib_array()` ）或返回值（ `fib_array_return_memory()` ）中的指针构建一个数组。Go 应用最终会调用 `deallocate` 来释放内存空间。
```
Comment (hydai) [inline docs/book/zh/src/embed/go/memory.md:365]: 中英文間應有空格隔開

**Review by hydai — CHANGES_REQUESTED** [review]

File: docs/book/zh/src/embed/go/memory.md (hunk 9 above)
Comment (Bevisy) [reply docs/book/zh/src/embed/go/memory.md]: Thanks, fixed now.

**Comment by Bevisy** [comment]
If there is nothing else that needs to be modified, I hope to be able to merge it as soon as possible, thank you! @hydai

**Review by hydai — APPROVED** [review]

**Pull request merged**